File: lib/block_encoding/pauli_lcu.cpp
/****************************************************************-*- C++ -*-****
 * Copyright (c) 2024 - 2025 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "cudaq/algorithms/block_encoding/pauli_lcu.h"
#include "cudaq/algorithms/block_encoding/kernels.h"
#include "cudaq/algorithms/detail/qpu_dispatch.h"
#include <cmath>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <string>
#include <utility>

namespace cudaq::algorithms {

// ============================================================================
// QUANTUM KERNELS (Stateless functors for CUDA-Q compiler)
// ============================================================================

__qpu__ void controlled_pauli_x(cudaq::qubit &control, cudaq::qview<> anc,
                                cudaq::qubit &target) {
  int n_anc = anc.size();
  CUDAQ_ALGORITHMS_APPLY_CONTROLLED_GATE_BY_ARITY(x, control, anc, target,
                                                  n_anc);
}

__qpu__ void controlled_pauli_y(cudaq::qubit &control, cudaq::qview<> anc,
                                cudaq::qubit &target) {
  int n_anc = anc.size();
  CUDAQ_ALGORITHMS_APPLY_CONTROLLED_GATE_BY_ARITY(y, control, anc, target,
                                                  n_anc);
}

__qpu__ void controlled_pauli_z(cudaq::qubit &control, cudaq::qview<> anc,
                                cudaq::qubit &target) {
  int n_anc = anc.size();
  CUDAQ_ALGORITHMS_APPLY_CONTROLLED_GATE_BY_ARITY(z, control, anc, target,
                                                  n_anc);
}

__qpu__ void controlled_sign_correction(cudaq::qubit &control,
                                        cudaq::qview<> anc) {
  int n_anc = anc.size();
  CUDAQ_ALGORITHMS_APPLY_CONTROLLED_Z_BY_ARITY(control, anc, n_anc);
}

/// @brief Kernel for externally controlled SELECT operation.
/// @note This intentionally mirrors block_encoding::controlled_select (in
/// device/kernels.cpp); the logic is kept duplicated rather than shared because
/// nvq++ cross-translation-unit device-kernel calls are fragile. If you change
/// the SELECT logic here, make the matching change in controlled_select (and
/// vice versa).
struct controlled_pauli_select_kernel {
  void operator()(cudaq::qubit &control, cudaq::qview<> anc, cudaq::qview<> sys,
                  const std::vector<int> &ctrls, const std::vector<int> &ops,
                  const std::vector<int> &lens,
                  const std::vector<int> &signs) const __qpu__ {
    int ptr_ctrl = 0;
    int ptr_op = 0;
    int n_anc = anc.size();

    for (std::size_t i = 0; i < lens.size(); ++i) {
      int n_ops = lens[i];
      int sign = signs[i];

      for (int b = 0; b < n_anc; ++b) {
        int bit_val = ctrls[ptr_ctrl++];
        if (bit_val == 0)
          x(anc[b]);
      }

      for (int k = 0; k < n_ops; ++k) {
        int code = ops[ptr_op++];
        int q_idx = ops[ptr_op++];

        if (code == 1)
          controlled_pauli_x(control, anc, sys[q_idx]);
        else if (code == 2)
          controlled_pauli_y(control, anc, sys[q_idx]);
        else if (code == 3)
          controlled_pauli_z(control, anc, sys[q_idx]);
      }

      if (sign < 0)
        controlled_sign_correction(control, anc);

      int back_ptr = ptr_ctrl - 1;
      for (int b_rev = 0; b_rev < n_anc; ++b_rev) {
        int anc_idx = (n_anc - 1) - b_rev;
        int bit_val = ctrls[back_ptr--];
        if (bit_val == 0)
          x(anc[anc_idx]);
      }
    }
  }
};

// ============================================================================
// CLASSICAL UTILITIES
// ============================================================================

namespace {

std::size_t ceil_log2(std::size_t value) {
  if (value <= 1)
    return 0;

  std::size_t power = 1;
  std::size_t log = 0;
  while (power < value) {
    power <<= 1;
    ++log;
  }
  return log;
}

bool is_identity_word(const cudaq::pauli_word &word) {
  const auto word_str = word.str();
  for (char pauli : word_str)
    if (pauli != 'I')
      return false;
  return true;
}

std::vector<double> compute_prepare_angles(const std::vector<double> &probs) {
  if (probs.empty())
    return {};

  std::size_t n_leaves = probs.size();
  if ((n_leaves & (n_leaves - 1)) != 0)
    throw std::runtime_error(
        "compute_prepare_angles: probability vector size must be a power of 2");

  int n_qubits = std::log2(n_leaves);
  std::vector<double> angles;

  // Build tree layer by layer
  for (int layer = 0; layer < n_qubits; ++layer) {
    int n_nodes = 1 << layer; // Number of nodes at this layer
    for (int node = 0; node < n_nodes; ++node) {
      int step = n_leaves / (1 << (layer + 1));
      int start_idx = node * step * 2;

      // Sum probabilities for current subtree
      double total_p = 0.0;
      for (int k = 0; k < 2 * step; ++k)
        total_p += probs[start_idx + k];

      // Guard against a zero-probability subtree, which occurs for the padded
      // leaves when num_terms is not a power of two (their probability is
      // exactly 0). This is a numerical zero-check, intentionally independent
      // of the user-facing `coefficient_threshold` used to drop small terms.
      constexpr double kZeroProbabilityTolerance = 1e-12;
      if (total_p < kZeroProbabilityTolerance) {
        angles.push_back(0.0);
      } else {
        // Sum of right branch
        double p_1 = 0.0;
        for (int k = step; k < 2 * step; ++k)
          p_1 += probs[start_idx + k];

        // Compute rotation angle: theta = 2*arcsin(sqrt(p_right / p_total))
        angles.push_back(2.0 * std::asin(std::sqrt(p_1 / total_p)));
      }
    }
  }

  return angles;
}

} // namespace

lcu_decomposition decompose_lcu(const cudaq::spin_op &hamiltonian,
                                std::size_t num_qubits,
                                double coefficient_threshold,
                                bool include_identity) {
  if (coefficient_threshold < 0.0)
    throw std::runtime_error(
        "decompose_lcu: coefficient threshold must be non-negative");

  lcu_decomposition lcu;
  lcu.num_system_qubits = num_qubits;
  lcu.coefficient_threshold = coefficient_threshold;
  lcu.include_identity = include_identity;

  for (const auto &term : hamiltonian) {
    auto coeff = term.evaluate_coefficient();
    double abs_coeff = std::abs(coeff);

    if (abs_coeff < coefficient_threshold)
      continue;

    if (std::abs(coeff.imag()) > 1e-10)
      throw std::runtime_error(
          "decompose_lcu: complex Hamiltonians not yet supported");

    auto word = term.get_pauli_word(num_qubits);
    double real_coeff = coeff.real();
    bool identity_term = is_identity_word(word);

    if (identity_term)
      lcu.constant_term += real_coeff;
    if (identity_term && !include_identity)
      continue;

    lcu.coefficients.push_back(real_coeff);
    lcu.absolute_coefficients.push_back(abs_coeff);
    lcu.signs.push_back((real_coeff < 0.0) ? -1 : 1);
    lcu.identity_terms.push_back(identity_term ? 1 : 0);
    lcu.pauli_words.push_back(word);
    lcu.normalization += abs_coeff;
  }

  lcu.num_terms = lcu.coefficients.size();
  if (lcu.num_terms == 0)
    throw std::runtime_error(
        include_identity
            ? "decompose_lcu: Hamiltonian has no retained terms"
            : "decompose_lcu: Hamiltonian has no retained non-identity terms");

  lcu.num_ancilla_qubits = ceil_log2(lcu.num_terms);
  if (lcu.num_ancilla_qubits >
      cudaq::algorithms::detail::max_lcu_ancilla_qubits)
    throw std::runtime_error(
        "decompose_lcu: PauliLCU currently supports at most " +
        std::to_string(cudaq::algorithms::detail::max_lcu_ancilla_qubits) +
        " ancilla qubits (" +
        std::to_string(1ULL
                       << cudaq::algorithms::detail::max_lcu_ancilla_qubits) +
        " LCU terms)");

  lcu.padded_num_terms = 1ULL << lcu.num_ancilla_qubits;

  for (double abs_coeff : lcu.absolute_coefficients)
    lcu.probabilities.push_back(abs_coeff / lcu.normalization);

  return lcu;
}

pauli_lcu_kernel_data make_pauli_lcu_kernel_data(const lcu_decomposition &lcu) {
  if (lcu.num_terms == 0)
    throw std::runtime_error(
        "make_pauli_lcu_kernel_data: LCU decomposition has no terms");
  if (lcu.coefficients.size() != lcu.num_terms ||
      lcu.pauli_words.size() != lcu.num_terms ||
      lcu.probabilities.size() != lcu.num_terms ||
      lcu.signs.size() != lcu.num_terms)
    throw std::runtime_error(
        "make_pauli_lcu_kernel_data: inconsistent LCU decomposition metadata");

  if (lcu.num_ancilla_qubits >
      cudaq::algorithms::detail::max_lcu_ancilla_qubits)
    throw std::runtime_error(
        "make_pauli_lcu_kernel_data: PauliLCU kernel data currently supports "
        "at most " +
        std::to_string(cudaq::algorithms::detail::max_lcu_ancilla_qubits) +
        " ancilla qubits");

  pauli_lcu_kernel_data data;
  data.num_system_qubits = lcu.num_system_qubits;
  data.num_terms = lcu.num_terms;
  data.padded_num_terms = lcu.padded_num_terms;
  data.num_ancilla_qubits = lcu.num_ancilla_qubits;

  auto padded_probabilities = lcu.probabilities;
  while (padded_probabilities.size() < lcu.padded_num_terms)
    padded_probabilities.push_back(0.0);

  data.state_prep_angles = compute_prepare_angles(padded_probabilities);

  for (std::size_t idx = 0; idx < lcu.num_terms; ++idx) {
    // Binary representation of index for control pattern
    for (std::size_t b = 0; b < data.num_ancilla_qubits; ++b) {
      // MSB to LSB ordering
      int bit_val = (idx >> (data.num_ancilla_qubits - 1 - b)) & 1;
      data.term_controls.push_back(bit_val);
    }

    // Parse Pauli word string: format is "XYZII..." where position i is qubit i
    std::string word_str = lcu.pauli_words[idx].str();
    std::vector<std::pair<int, int>> ops_for_term; // (code, qubit_idx)

    // Each character position corresponds to a qubit
    for (std::size_t q_idx = 0; q_idx < word_str.size(); ++q_idx) {
      char pauli_char = word_str[q_idx];

      if (pauli_char == 'I') {
        // Identity - skip
        continue;
      } else if (pauli_char == 'X') {
        ops_for_term.push_back({1, static_cast<int>(q_idx)});
      } else if (pauli_char == 'Y') {
        ops_for_term.push_back({2, static_cast<int>(q_idx)});
      } else if (pauli_char == 'Z') {
        ops_for_term.push_back({3, static_cast<int>(q_idx)});
      }
      // Other characters are ignored
    }

    // Flatten ops
    for (const auto &[code, q_idx] : ops_for_term) {
      data.term_ops.push_back(code);
      data.term_ops.push_back(q_idx);
    }

    data.term_lengths.push_back(static_cast<int>(ops_for_term.size()));
    data.term_signs.push_back(lcu.signs[idx]);
  }

  return data;
}

std::vector<double>
pauli_lcu::compute_angles(const std::vector<double> &probs) {
  return compute_prepare_angles(probs);
}

// ============================================================================
// PAULI LCU IMPLEMENTATION
// ============================================================================

pauli_lcu::pauli_lcu(const cudaq::spin_op &hamiltonian, std::size_t num_qubits)
    : pauli_lcu(hamiltonian, num_qubits, true) {}

pauli_lcu::pauli_lcu(const cudaq::spin_op &hamiltonian, std::size_t num_qubits,
                     bool include_identity)
    : pauli_lcu(
          decompose_lcu(hamiltonian, num_qubits, 1e-12, include_identity)) {}

pauli_lcu::pauli_lcu(const lcu_decomposition &lcu)
    : kernel_data(make_pauli_lcu_kernel_data(lcu)), decomposition(lcu) {
  n_sys = kernel_data.num_system_qubits;
  n_anc = kernel_data.num_ancilla_qubits;
  alpha = decomposition.normalization;
}

// ============================================================================
// QUANTUM OPERATION DISPATCHERS
// ============================================================================

void pauli_lcu::prepare(cudaq::qview<> ancilla) const {
  if (ancilla.size() != n_anc)
    throw std::runtime_error("pauli_lcu::prepare: ancilla size mismatch");
  block_encoding::prepare(ancilla, kernel_data.state_prep_angles);
}

void pauli_lcu::unprepare(cudaq::qview<> ancilla) const {
  if (ancilla.size() != n_anc)
    throw std::runtime_error("pauli_lcu::unprepare: ancilla size mismatch");
  block_encoding::unprepare(ancilla, kernel_data.state_prep_angles);
}

void pauli_lcu::select(cudaq::qview<> ancilla, cudaq::qview<> system) const {
  if (ancilla.size() != n_anc)
    throw std::runtime_error("pauli_lcu::select: ancilla size mismatch");
  if (system.size() != n_sys)
    throw std::runtime_error("pauli_lcu::select: system size mismatch");

  block_encoding::select(ancilla, system, kernel_data.term_controls,
                         kernel_data.term_ops, kernel_data.term_lengths,
                         kernel_data.term_signs);
}

void pauli_lcu::controlled_select(cudaq::qubit &control, cudaq::qview<> ancilla,
                                  cudaq::qview<> system) const {
  if (ancilla.size() != n_anc)
    throw std::runtime_error(
        "pauli_lcu::controlled_select: ancilla size mismatch");
  if (system.size() != n_sys)
    throw std::runtime_error(
        "pauli_lcu::controlled_select: system size mismatch");

  controlled_pauli_select_kernel{}(
      control, ancilla, system, kernel_data.term_controls, kernel_data.term_ops,
      kernel_data.term_lengths, kernel_data.term_signs);
}

} // namespace cudaq::algorithms

File: experiments/lcu_python/cudaq_algorithms/pauli_lcu.py
# ============================================================================ #
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
"""Pauli LCU block encoding.

Implements the linear-combination-of-unitaries block encoding of a Pauli-sum
Hamiltonian: the LCU decomposition, PREPARE-angle computation, and the
PREPARE/SELECT/walk device kernels (including controlled variants).

``PauliLCU`` is the primary entry point. It accepts a ``cudaq.SpinOperator``,
a ``{"XZI...": coeff}`` mapping, or ``[(coeff, word), ...]`` pairs, and
provides ready-to-run kernel factories::

    enc = PauliLCU({"ZI": 0.7, "XX": 0.19, "IZ": -0.43})
    kernel = enc.encode_kernel()            # @cudaq.kernel(state)

The module-level kernels (``prepare``, ``select``, ``apply``, ...) are
composable from user kernels; ``PauliLCU.kernel_args`` supplies the flattened
arrays they take as arguments. Statevector-based conveniences (postselection,
``action``) live in ``sim_utils``, outside the library surface.

The encoding satisfies ``(<0|_anc ⊗ I) U (|0>_anc ⊗ I) = H / alpha`` with
``alpha = sum(|c_i|)``, and the qubitization walk block is ``-H / alpha``
(the zero reflection phases ``|0...0>`` by -1).
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Mapping

import cudaq

_PAULI_CODES = {"X": 1, "Y": 2, "Z": 3}


# ============================================================================
# Device kernels (module level, composable from user kernels)
# ============================================================================


@cudaq.kernel
def prepare(ancilla: cudaq.qview, angles: list[float]):
    """PREPARE: encode sqrt(|c_i| / alpha) amplitudes on the ancilla register."""
    n = ancilla.size()
    if n == 0:
        return
    ry(angles[0], ancilla[0])
    idx = 1
    for layer in range(1, n):
        branches = 1 << layer
        for branch in range(branches):
            for bit in range(layer):
                if ((branch >> bit) & 1) == 0:
                    x(ancilla[layer - 1 - bit])
            ry.ctrl(angles[idx], ancilla.front(layer), ancilla[layer])
            idx += 1
            for bit in range(layer):
                if ((branch >> bit) & 1) == 0:
                    x(ancilla[layer - 1 - bit])


@cudaq.kernel
def unprepare(ancilla: cudaq.qview, angles: list[float]):
    """PREPARE dagger."""
    n = ancilla.size()
    if n == 0:
        return
    idx = len(angles) - 1
    for reverse_layer in range(n - 1):
        layer = n - 1 - reverse_layer
        branches = 1 << layer
        for reverse_branch in range(branches):
            branch = branches - 1 - reverse_branch
            for bit in range(layer):
                if ((branch >> bit) & 1) == 0:
                    x(ancilla[layer - 1 - bit])
            ry.ctrl(-angles[idx], ancilla.front(layer), ancilla[layer])
            idx -= 1
            for bit in range(layer):
                if ((branch >> bit) & 1) == 0:
                    x(ancilla[layer - 1 - bit])
    ry(-angles[0], ancilla[0])


@cudaq.kernel
def select(ancilla: cudaq.qview, system: cudaq.qview, term_controls: list[int],
           term_ops: list[int], term_lengths: list[int],
           term_signs: list[int]):
    """SELECT: apply Pauli term i, controlled on ancilla state |i>."""
    n_anc = ancilla.size()
    ptr_ctrl = 0
    ptr_op = 0
    for i in range(len(term_lengths)):
        for b in range(n_anc):
            if term_controls[ptr_ctrl] == 0:
                x(ancilla[b])
            ptr_ctrl += 1
        for _ in range(term_lengths[i]):
            code = term_ops[ptr_op]
            target = term_ops[ptr_op + 1]
            ptr_op += 2
            if n_anc == 0:
                if code == 1:
                    x(system[target])
                elif code == 2:
                    y(system[target])
                else:
                    z(system[target])
            else:
                if code == 1:
                    x.ctrl(ancilla, system[target])
                elif code == 2:
                    y.ctrl(ancilla, system[target])
                else:
                    z.ctrl(ancilla, system[target])
        if term_signs[i] < 0:
            if n_anc == 0:
                # Single-term LCU: no projected block exists, so the sign is
                # part of the encoded operator. rz(2*pi) is exactly -I.
                rz(6.283185307179586, system[0])
            elif n_anc == 1:
                z(ancilla[0])
            else:
                z.ctrl(ancilla.front(n_anc - 1), ancilla[n_anc - 1])
        back = ptr_ctrl - 1
        for b in range(n_anc):
            if term_controls[back] == 0:
                x(ancilla[n_anc - 1 - b])
            back -= 1


@cudaq.kernel
def controlled_select(control_and_ancilla: cudaq.qview, system: cudaq.qview,
                      term_controls: list[int], term_ops: list[int],
                      term_lengths: list[int], term_signs: list[int]):
    """SELECT controlled by an external qubit.

    CUDA-Q Python kernels cannot mix a bare qubit with a qview in one
    control set, so the external control is qubit 0 of
    ``control_and_ancilla`` and the LCU ancillas are the remaining qubits;
    every control set is then a view of that one register.
    """
    n_anc = control_and_ancilla.size() - 1
    ptr_ctrl = 0
    ptr_op = 0
    for i in range(len(term_lengths)):
        for b in range(n_anc):
            if term_controls[ptr_ctrl] == 0:
                x(control_and_ancilla[1 + b])
            ptr_ctrl += 1
        for _ in range(term_lengths[i]):
            code = term_ops[ptr_op]
            target = term_ops[ptr_op + 1]
            ptr_op += 2
            if code == 1:
                x.ctrl(control_and_ancilla, system[target])
            elif code == 2:
                y.ctrl(control_and_ancilla, system[target])
            else:
                z.ctrl(control_and_ancilla, system[target])
        if term_signs[i] < 0:
            if n_anc == 0:
                # Controlled -I on the system is a Z on the control.
                z(control_and_ancilla[0])
            else:
                total = control_and_ancilla.size()
                z.ctrl(control_and_ancilla.front(total - 1),
                       control_and_ancilla[total - 1])
        back = ptr_ctrl - 1
        for b in range(n_anc):
            if term_controls[back] == 0:
                x(control_and_ancilla[n_anc - b])
            back -= 1


@cudaq.kernel
def apply(ancilla: cudaq.qview, system: cudaq.qview, angles: list[float],
          term_controls: list[int], term_ops: list[int],
          term_lengths: list[int], term_signs: list[int]):
    """Full block encoding: PREPARE, SELECT, PREPARE dagger."""
    prepare(ancilla, angles)
    select(ancilla, system, term_controls, term_ops, term_lengths, term_signs)
    unprepare(ancilla, angles)


@cudaq.kernel
def reflect_about_zero(register: cudaq.qview):
    """I - 2|0...0><0...0| (phases the all-zero state by -1)."""
    n = register.size()
    if n == 0:
        return
    for i in range(n):
        x(register[i])
    if n == 1:
        z(register[0])
    else:
        z.ctrl(register.front(n - 1), register[n - 1])
    for i in range(n):
        x(register[i])


@cudaq.kernel
def walk(ancilla: cudaq.qview, system: cudaq.qview, angles: list[float],
         term_controls: list[int], term_ops: list[int],
         term_lengths: list[int], term_signs: list[int]):
    """One qubitization walk step: SELECT, then reflect about PREPARE."""
    select(ancilla, system, term_controls, term_ops, term_lengths, term_signs)
    unprepare(ancilla, angles)
    reflect_about_zero(ancilla)
    prepare(ancilla, angles)


def state_from(ket):
    """Build a cudaq.State from array data at the current target's precision.

    fp32 simulators (e.g. the default `nvidia` target) reject complex128
    input ("[sim-state] invalid data precision"); cudaq.complex() reports
    the dtype the active target expects.
    """
    import numpy as np

    return cudaq.State.from_data(np.asarray(ket, dtype=cudaq.complex()))


# ============================================================================
# Host-side decomposition
# ============================================================================


def _terms_from_input(hamiltonian, num_qubits):
    """Normalize the supported Hamiltonian input forms to (coeff, word) pairs."""
    if isinstance(hamiltonian, Mapping):
        pairs = [(float(c), str(w)) for w, c in hamiltonian.items()]
    elif isinstance(hamiltonian, cudaq.SpinOperator):
        width = num_qubits if num_qubits is not None else int(
            hamiltonian.qubit_count)
        pairs = []
        for term in hamiltonian:
            coeff = term.evaluate_coefficient()
            if abs(coeff.imag) > 1e-10:
                raise ValueError(
                    "complex Hamiltonian coefficients are not supported")
            pairs.append((float(coeff.real), str(term.get_pauli_word(width))))
    elif isinstance(hamiltonian, Iterable):
        pairs = [(float(c), str(w)) for c, w in hamiltonian]
    else:
        raise TypeError(
            "hamiltonian must be a cudaq.SpinOperator, a {word: coeff} "
            "mapping, or an iterable of (coeff, word) pairs")

    if not pairs:
        raise ValueError("hamiltonian has no terms")

    width = len(pairs[0][1])
    for _, word in pairs:
        if len(word) != width:
            raise ValueError("all Pauli words must have the same length")
        if any(ch not in "IXYZ" for ch in word):
            raise ValueError(f"unsupported Pauli word: {word!r}")
    if num_qubits is not None and num_qubits != width:
        raise ValueError(
            f"num_qubits={num_qubits} does not match word length {width}")
    return pairs, width


def _prepare_angles(probabilities):
    """Rotation angles for the binary state-preparation tree."""
    n_leaves = len(probabilities)
    if n_leaves & (n_leaves - 1):
        raise ValueError("probability vector size must be a power of 2")
    n_qubits = n_leaves.bit_length() - 1

    angles = []
    for layer in range(n_qubits):
        step = n_leaves >> (layer + 1)
        for node in range(1 << layer):
            start = node * step * 2
            total = sum(probabilities[start:start + 2 * step])
            if total < 1e-12:
                # Zero-probability subtree from power-of-two padding.
                angles.append(0.0)
            else:
                right = sum(probabilities[start + step:start + 2 * step])
                angles.append(2.0 * math.asin(math.sqrt(right / total)))
    return angles


# ============================================================================
# The user-facing object
# ============================================================================


class PauliLCU:
    """Block encoding of H / alpha via a linear combination of Pauli strings.

    Parameters
    ----------
    hamiltonian
        A ``cudaq.SpinOperator``, a mapping ``{"XZI...": coefficient}``, or an
        iterable of ``(coefficient, word)`` pairs. Words use one I/X/Y/Z
        character per qubit, position = qubit index.
    num_qubits
        Optional; inferred from the input when omitted, validated against it
        when given.
    include_identity
        Whether identity words are retained inside the encoded operator
        (their sum is always reported as ``constant_term``).
    coefficient_threshold
        Terms with ``|coefficient|`` below this are dropped.
    """

    def __init__(self, hamiltonian, num_qubits=None, *, include_identity=True,
                 coefficient_threshold=1e-12):
        pairs, width = _terms_from_input(hamiltonian, num_qubits)

        self._num_system = width
        self._constant = 0.0
        kept = []
        for coeff, word in pairs:
            if abs(coeff) < coefficient_threshold:
                continue
            identity = set(word) == {"I"}
            if identity:
                self._constant += coeff
                if not include_identity:
                    continue
            kept.append((coeff, word))
        if not kept:
            raise ValueError("hamiltonian has no retained terms")

        self._terms = kept
        self._alpha = sum(abs(c) for c, _ in kept)
        self._num_ancilla = max(0, (len(kept) - 1).bit_length())

        # Precompute the flattened arrays that cross the kernel boundary.
        probabilities = [abs(c) / self._alpha for c, _ in kept]
        probabilities += [0.0] * ((1 << self._num_ancilla) - len(kept))
        self._angles = _prepare_angles(probabilities)

        self._term_controls = []
        self._term_ops = []
        self._term_lengths = []
        self._term_signs = []
        for index, (coeff, word) in enumerate(kept):
            for b in range(self._num_ancilla):
                self._term_controls.append(
                    (index >> (self._num_ancilla - 1 - b)) & 1)
            ops = [(code, qubit) for qubit, ch in enumerate(word)
                   if (code := _PAULI_CODES.get(ch)) is not None]
            for code, qubit in ops:
                self._term_ops.extend((code, qubit))
            self._term_lengths.append(len(ops))
            self._term_signs.append(-1 if coeff < 0.0 else 1)

    # ------------------------------------------------------------------
    # Inspection
    # ------------------------------------------------------------------

    @property
    def num_system(self) -> int:
        return self._num_system

    @property
    def num_ancilla(self) -> int:
        return self._num_ancilla

    @property
    def num_terms(self) -> int:
        return len(self._terms)

    @property
    def alpha(self) -> float:
        """LCU normalization (1-norm of retained coefficients)."""
        return self._alpha

    normalization = alpha

    @property
    def constant_term(self) -> float:
        return self._constant

    @property
    def terms(self) -> list[tuple[float, str]]:
        return list(self._terms)

    @property
    def kernel_args(self):
        """(angles, term_controls, term_ops, term_lengths, term_signs).

        Escape hatch for composing the module-level kernels inside your own
        ``@cudaq.kernel``; the factory methods below capture these for you.
        """
        return (list(self._angles), list(self._term_controls),
                list(self._term_ops), list(self._term_lengths),
                list(self._term_signs))

    def __repr__(self):
        return (f"PauliLCU(terms={self.num_terms}, "
                f"system_qubits={self.num_system}, "
                f"ancilla_qubits={self.num_ancilla}, "
                f"alpha={self.alpha:.6g})")

    # ------------------------------------------------------------------
    # Kernel factories
    # ------------------------------------------------------------------

    def _zero_ancilla_kernel(self, repetitions):
        """Single-term (0-ancilla) special case.

        Needed because CUDA-Q cannot marshal empty ``list`` kernel arguments
        ("Cannot infer runtime argument type"), and with zero ancillas the
        flattened control data is empty. The encoding degenerates to the
        signed Pauli word, so apply it directly. One walk step also reduces
        to one signed application (the reflections are empty-register no-ops).
        """
        ops = list(self._term_ops)
        negative = self._term_signs[0] < 0
        steps = int(repetitions)

        if ops:

            @cudaq.kernel
            def single_term(state: cudaq.State):
                system = cudaq.qvector(state)
                for _ in range(steps):
                    for i in range(len(ops) // 2):
                        code = ops[2 * i]
                        target = ops[2 * i + 1]
                        if code == 1:
                            x(system[target])
                        elif code == 2:
                            y(system[target])
                        else:
                            z(system[target])
                    if negative:
                        rz(6.283185307179586, system[0])

            return single_term

        @cudaq.kernel
        def single_identity_term(state: cudaq.State):
            system = cudaq.qvector(state)
            for _ in range(steps):
                if negative:
                    rz(6.283185307179586, system[0])

        return single_identity_term

    def encode_kernel(self):
        """A ``@cudaq.kernel(state)`` applying the full block encoding.

        The kernel allocates the system register from ``state`` and the
        ancilla register (in |0...0>) after it.
        """
        if self.num_ancilla == 0:
            return self._zero_ancilla_kernel(1)

        angles, controls, ops, lengths, signs = self.kernel_args
        n_anc = self.num_ancilla

        @cudaq.kernel
        def encoded(state: cudaq.State):
            system = cudaq.qvector(state)
            ancilla = cudaq.qvector(n_anc)
            apply(ancilla, system, angles, controls, ops, lengths, signs)

        return encoded

    def walk_kernel(self, power: int = 1):
        """A ``@cudaq.kernel(state)`` running PREPARE, walk^power, UNPREPARE.

        The all-zero-ancilla block of the result is T_power(-H/alpha) applied
        to the input state (Chebyshev polynomial of the walk block -H/alpha).
        """
        if self.num_ancilla == 0:
            return self._zero_ancilla_kernel(power)

        angles, controls, ops, lengths, signs = self.kernel_args
        n_anc = self.num_ancilla
        steps = int(power)

        @cudaq.kernel
        def walked(state: cudaq.State):
            system = cudaq.qvector(state)
            ancilla = cudaq.qvector(n_anc)
            prepare(ancilla, angles)
            for _ in range(steps):
                walk(ancilla, system, angles, controls, ops, lengths, signs)
            unprepare(ancilla, angles)

        return walked

